File: oscar/src/edu/virginia/engine/display/DisplayObject.js
"use strict";

/**
 * A very basic display object for a javascript based gaming engine
 *
 * */
class DisplayObject extends IEventDispatcher{

	constructor(id, filename){
		super();
		this.id = id;
		this.displayImage = new Image();
		this.loaded = false;
		this.loadImage(filename);
		this.parent = null; //new DisplayObject();
		// Transformation variables
		this.visible = true;
		this.xPos = 0;
		this.yPos = 0;
		this.xPivot = 0;
		this.yPivot = 0;
		this.scaleX = 1.0;
		this.scaleY = 1.0;
		this.rotation = 0;
		this.alpha = 1.0;
		// Transformation booleans for toggles
		this.visibleHeld = false;
		this.inPlay = true;
	}

	/**
	 * Loads the image, sets a flag called 'loaded' when the image is ready to be drawn
	 */
	loadImage(filename){
		// Get file, if it exists
		if (filename != null){
			try {
	    	this.displayImage.src = 'resources/' + filename;
				this.loaded = true;
			}
			catch(err) {
				alert("Incorrect Filename: " + filename);
			}
		}
	}

	/**
	 * Invoked every frame, manually for now, but later automatically if this DO is in DisplayTree
	 */
	update(pressedKeys){
	}

	/**
	 * Draws this image to the screen
	 */
	draw(g){
		if(this.displayImage){
			var oldSelf = this.applyTransformations(g);
			if(this.loaded && this.visible){
				g.drawImage(this.displayImage,this.xPos,this.yPos);
			}
			this.reverseTransformations(g, oldSelf);
		}
	}

	/* For debugging */
	printSelf(){
		console.log(this.visible);
		console.log(this.xPos);
		console.log(this.yPos);
		console.log(this.xPivot);
		console.log(this.yPivot);
		console.log(this.scaleX);
		console.log(this.scaleY);
		console.log(this.rotation);
		console.log(this.alpha);
		console.log("~~~~~~~~~~~~~~~~~");
	}

	/* Save old values before apply transformations */
	getOldSelf(){
		var oldSelf = new ArrayList();
		oldSelf.push(this.visible);
		oldSelf.push(this.xPos);
		oldSelf.push(this.yPos);
		oldSelf.push(this.xPivot);
		oldSelf.push(this.yPivot);
		oldSelf.push(this.scaleX);
		oldSelf.push(this.scaleY);
		oldSelf.push(this.rotation);
		oldSelf.push(this.alpha);
		return oldSelf;
	}

	/* Restore values before apply transformations */
	restoreSelf(oldSelf){
		this.visible = oldSelf.get(0);
		this.xPos = oldSelf.get(1);
		this.yPos = oldSelf.get(2);
		this.xPivot = oldSelf.get(3);
		this.yPivot = oldSelf.get(4);
		this.scaleX = oldSelf.get(5);
		this.scaleY = oldSelf.get(6);
		this.rotation = oldSelf.get(7);
		this.alpha = oldSelf.get(8);
	}


	/* Create a temporary child based on its parent attributes */
	getParentAttributes(parent){
		// Another parent exists, keep using recursion
		if (parent.parent != null){
			this.getParentAttributes(parent.parent);
		}
		/* Parent's pivot point is considered 0,0 of sprite's universe */
		this.xPos += parent.xPivot;
		this.yPos += parent.yPivot;
		/* To find out your parent's global pivot point */
		this.xPivot += parent.xPivot;
		this.yPivot += parent.yPivot;
		/* Scale by mutiplication */
		this.scaleX *= parent.scaleX;
		this.scaleY *= parent.scaleY;
	}

	/**
	 * Applies transformations for this display object to the given graphics
	 * object
	 * */
	applyTransformations(g) {
		g.save(); // ends with g.restore() in reverseTransformations()
		var oldSelf = this.getOldSelf();
		// rotate/scale around parent
		if (this.parent != null){
			this.getParentAttributes(this.parent);
			/* Lab Six Game is root, no attributes to obtain */
			if (this.parent.id == "Lab Six Game"){
				var xPivot = this.xPos;
				var yPivot = this.yPos;
			}
			/* Pivot around the global x,y position of your parent pivots,
			 * which does NOT include your personal pivot point */
			else {
				 var xPivot = this.xPivot - oldSelf.get(3);
				 var yPivot = this.yPivot - oldSelf.get(4);
			}

			g.translate(xPivot, yPivot); //translate to pivot point
			g.rotate(this.rotation*Math.PI/180); // rotation
			g.scale(this.scaleX, this.scaleY); // grow/shrink
			g.translate(-xPivot, -yPivot); //translate back to origin
			// increase/decrease alpha (not needed for lab)
			g.globalAlpha = this.alpha;
		}
		/* Needed for reverse Transformations */
		return oldSelf;
	}

	/**
	 * Reverses transformations for this display object to the given graphics
	 * object
	 * */
	reverseTransformations(g, oldSelf) {
		g.restore(); //to end g.save() in applyTransformations()
		/* Restore old variable values */
		//this.printSelf();
		this.restoreSelf(oldSelf);
	}

	/* For collision detection,
		gets global coordinates of DisplayObject */
	getHitbox(parent, hitbox){
		var root = false;
		if (hitbox == null){
			var hitbox = new Map();
			hitbox.set('left', this.xPos);
			hitbox.set('top', this.yPos);
			root = true;
		}
		/* Another parent exists, keep using recursion */
		if (parent.parent != null){
			hitbox = this.getHitbox(parent.parent);
		}

		/* Variables used by both conditionals */
		var left = hitbox.get('left');
		var top = hitbox.get('top');
		/* Return final hitbox coordinates */
		if (root){
			hitbox.set('right', left + this.getScaledWidth());
			hitbox.set('bottom', top + this.getScaledHeight());
			return hitbox;
		}
		/* Continue discovering global coordinates */
		else {
			/* Parent's pivot point is considered 0,0 of sprite's universe */
			hitbox.set('left', left + parent.xPivot);
			hitbox.set('top', top + parent.yPivot);
			return hitbox;
		}
	}
	/* Boolean for rectangle collision */
	collidesWith(other){
		/* My boundaries */
		var mhb = this.getHitbox(this.parent); //myHitbox
		var aLeft = mhb.get('left');
		var aRight = mhb.get('right');
		var aTop = mhb.get('top');
		var aBottom = mhb.get('bottom');
		/* Other Boundaries*/
		var ohb = other.getHitbox(other.parent); //otherHitBox
		var bLeft = ohb.get('left');
		var bRight = ohb.get('right');
		var bTop = ohb.get('top');
		var bBottom = ohb.get('bottom');
		/* Dispatch events on collisions  */
		if (aLeft <= bRight && bLeft <= aRight && aTop <= bBottom && bTop <= aBottom){
				if(other.id == "Sun"){
					var pickup = new Event('Pickup', other);
					this.dispatchEvent(pickup);
					this.inPlay = false;
				}
				/* Generic collisions */
				else{
					var collision = new Event('Collision', other);
					this.dispatchEvent(pickup);
				}
		}
	}

	/* Boolean for rectangle collision */
	isTouching(other){
		/* My boundaries */
		var aLeft = this.xPos;
		var aRight = this.xPos + this.getScaledWidth();
		var aTop = this.yPos;
		var aBottom = this.yPos + this.getScaledHeight();
		/* Other Boundaries*/
		var bLeft = other.xPos;
		var bRight = other.xPos + other.getScaledWidth();
		var bTop = other.yPos;
		var bBottom = other.yPos + other.getScaledHeight();
		/* Collision bools */
		return (aLeft <= bRight &&
          bLeft <= aRight &&
          aTop <= bBottom &&
          bTop <= aBottom);
	}

	/**
	 * THIS AREA CONTAINS MOSTLY GETTERS AND SETTERS!
	 *
	 */

	setId(id){this.id = id;}
	getId(){return this.id;}

	setDisplayImage(image){this.displayImage = image;} //image needs to already be loaded!
	getDisplayImage(){return this.displayImage;}

	getUnscaledHeight(){return this.displayImage.height;}
	getUnscaledWidth(){return this.displayImage.width;}

	getScaledHeight(){
		var unscaledHeight = this.getUnscaledHeight();
		return unscaledHeight*this.scaleY;
	}
	getScaledWidth(){
		var unscaledWidth = this.getUnscaledWidth();
		return unscaledWidth*this.scaleX;
	}

	/* Getters and Setters */
	// Position
	getPositionX(){return this.xPos;}
	getPositionY(){return this.yPos;}
	setPositionX(xPos){this.xPos = xPos;}
	setPositionY(yPos){this.yPos = yPos;}
	setPosition(xPos, yPos){
		this.xPos = xPos;
		this.yPos = yPos;
	}

	setEllipseRadius(base, extension){
		this.xPos = base + Math.abs(extension*Math.cos(this.rotation*Math.PI/180));
		if (this.xPos - base < 10){this.xPos = base+10;}
	}

	// Pivot Point
	getPivotPointX(){return this.xPivot;}
	getPivotPointY(){return this.yPivot;}
	setPivotPointX(xPivot){this.xPivot = xPivot;}
	setPivotPointY(yPivot){this.yPivot = yPivot;}

	// ScaleX
	getScaleX(){return this.scaleX;}
	setScaleX(scaleX){
		var limitX = scaleX;
		if (scaleX > 2.0){limitX = 2.0;}
		if (scaleX < 0.1){limitX = 0.1;}
		this.scaleX = limitX;
	}

	// ScaleX
	getScaleY(){return this.scaleY;}
	setScaleY(scaleY){
		var limitY = scaleY;
		if (scaleY > 2.0){limitY = 2.0;}
		if (scaleY < 0.1){limitY = 0.1;}
		this.scaleY = limitY;
	}

	// Rotation
	getRotation(){return this.rotaion;}
	setRotation(newRotation){
		var limitRotation = currentRotation;
		/* No angles beyond 0 or 360 degrees */
		if (limitRotation < 0){limitRotation += 360;}
		if (limitRotation > 360){limitRotation -= 360;}
		this.rotation = limitRotation;
	}
	addRotation(newRotation){
		var currentRotation = this.rotation;
		var limitRotation = currentRotation + newRotation;
		/* No angles beyond 0 or 360 degrees */
		if (limitRotation < 0){limitRotation += 360;}
		if (limitRotation > 360){limitRotation -= 360;}
		this.rotation = limitRotation;
	}

	// Alpha
	getAlpha(){return this.alpha;}
	setAlpha(alpha){
		var newAlpha = alpha;
		if (newAlpha < 0){newAlpha = 0;}
		else if (newAlpha > 1){newAlpha = 1;}
		this.alpha = newAlpha;
	}
}

File: app/src/edu/virginia/engine/display/DisplayObject.java
package edu.virginia.engine.display;

import java.awt.AlphaComposite;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

import javax.imageio.ImageIO;

import edu.virginia.engine.util.EventDispatcher;

/**
 * A very basic display object for a java based gaming engine
 * 
 * */
public class DisplayObject extends EventDispatcher{

	/* All DisplayObject have a unique id */
	private String id;

	/* The image that is displayed by this object */
	private BufferedImage displayImage;
	
	private boolean visible;
	private boolean inPlay = true;
	private Point position;
	private Point pivotPoint;
	private double scaleX;
	private double scaleY;
	private double rotation;
	private float alpha;
	private Rectangle hitbox;
	
	private DisplayObject parent;
	
	
	

	/**
	 * Constructors: can pass in the id OR the id and image's file path and
	 * position OR the id and a buffered image and position
	 */
	public DisplayObject(String id) {
		this.setId(id);
		this.visible=true;
		this.position=new Point(0,0);
		this.pivotPoint=new Point(0,0);
		this.scaleX=1;
		this.scaleY=1;
		this.rotation=0;
		this.alpha=1;
		//this.addEventListener(myQuestManager, CollisionEvent.COLLISION);
	}

	public DisplayObject(String id, String fileName) {
		this.setId(id);
		this.setImage(fileName);
		this.visible=true;
		this.position=new Point(0,0);
		this.pivotPoint=new Point(0,0);
		this.scaleX=1;
		this.scaleY=1;
		this.rotation=0;
		this.alpha=1;
		//this.addEventListener(myQuestManager, CollisionEvent.COLLISION);
	}

	
	//Getters, Setters___________________________________________________________________________________
	public void setId(String id) {
		this.id = id;
	}

	public String getId() {
		return id;
	}


	/**
	 * Returns the unscaled width and height of this display object
	 * */
	public int getUnscaledWidth() {
		if(displayImage == null) return 0;
		return displayImage.getWidth();
	}

	public int getUnscaledHeight() {
		if(displayImage == null) return 0;
		return displayImage.getHeight();
	}

	public BufferedImage getDisplayImage() {
		return this.displayImage;
	}

	public void setImage(String imageName) {
		if (imageName == null) {
			return;
		}
		displayImage = readImage(imageName);
		if (displayImage == null) {
			System.err.println("[DisplayObject.setImage] ERROR: " + imageName + " does not exist!");
		}
	}
	
	public boolean getVisible(){
		return this.visible;
	}
	
	public void setVisible(boolean b){
		this.visible=b;
	}
	
	public Point getPosition(){
		return this.position;
	}
	
	public void setPosition(Point p){
		this.position=p;
	}
	
	public Point getGlobalPosition(){
		int parentX = 0;
		int parentY = 0;
		// Parent exists, use recursion to find global coordinates
		if (this.parent != null){
			Point parentCoor = this.parent.getGlobalPosition();
			parentX = (int) parentCoor.getX();
			parentY = (int) parentCoor.getY();
		}
		int ourX = (int) this.getPosition().getX();
		int ourY = (int) this.getPosition().getY();
		int currX = ourX + parentX;
		int currY = ourY + parentY;
		if (this.getId().equals("Platform2")){System.out.format("Plat2 x: %d y: %d\n",currX, currY);}
		return (new Point(currX, currY));
	}
	
	public Point getPivotPoint(){
		return this.pivotPoint;
	}
	
	public void setPivotPoint(Point p){
		this.pivotPoint=p;
	}
	
	public double getScaleX(){
		return this.scaleX;
	}
	
	public void setScaleX(double d){
		this.scaleX=d;
	}

	public double getScaleY(){
		return this.scaleY;
	}
	
	public void setScaleY(double d){
		this.scaleY=d;
	}
	public double getRotation(){
		return this.rotation;
	}
	
	public void setRotation(double i){
		this.rotation=i%360;
	}
	
	public float getAlpha(){
		return this.alpha;
	}
	
	public void setAlpha(float i){
		this.alpha=i;
	}
	
	public Rectangle getHitbox(){
		Rectangle hitbox = new Rectangle((int)this.position.getX(),(int)this.position.getY(),this.getUnscaledWidth(), this.getUnscaledHeight());
		return hitbox;
	}
	
	public DisplayObject getParent(){
		if (this.parent != null){
			return this.parent;
		}
		return null;
	}
	
	public void setParent (DisplayObject o){
		this.parent=o;
	}
	//___________________________________________________________________________________________________-
	
	public boolean collidesWith(DisplayObject other){
		if (this.getHitbox().intersects(other.getHitbox())){
			return true;
		}
		return false;
	}
	
	public boolean collidesWithRectangle(Rectangle other){
		if (this.getHitbox().intersects(other)){
			return true;
		}
		return false;
	}
	
	public int getCenterX(){
		return (int) this.getPosition().getX() + this.getUnscaledWidth()/2;
	}
	
	public int getCenterY(){
		return (int) this.getPosition().getY() + this.getUnscaledHeight()/2;
	}
	
	public float getAngle(Point target) {
		/* x and y are both the center of the DisplayObject */
		int x = this.getCenterX();
		int y = this.getCenterY();
		
	    float angle = (float) Math.toDegrees(Math.atan2(target.y - y, target.x - x));

	    if(angle < 0){
	        angle += 360;
	    }
	    //System.out.println("Angle: " + angle);
	    return angle;
	}
	
	/**
	 * Helper function that simply reads an image from the given image name
	 * (looks in resources\\) and returns the bufferedimage for that filename
	 * */
	public BufferedImage readImage(String imageName) {
		BufferedImage image = null;
		try {
			String file = ("resources" + File.separator + imageName);
			image = ImageIO.read(new File(file));
		} catch (IOException e) {
			System.out.println("[Error in DisplayObject.java:readImage] Could not read image " + imageName);
			e.printStackTrace();
		}
		return image;
	}

	public void setImage(BufferedImage image) {
		if(image == null) return;
		displayImage = image;
	}


	/**
	 * Invoked on every frame before drawing. Used to update this display
	 * objects state before the draw occurs. Should be overridden if necessary
	 * to update objects appropriately.
	 * */
	protected void update(ArrayList<Integer> pressedKeys) {
	}

	/**
	 * Draws this image. This should be overloaded if a display object should
	 * draw to the screen differently. This method is automatically invoked on
	 * every frame.
	 * */
	//from http://www.informit.com/articles/article.aspx?p=26349&seqNum=5
	 private AlphaComposite makeComposite(float alpha) {
		  int type = AlphaComposite.SRC_OVER;
		  return(AlphaComposite.getInstance(type, alpha));
		 }
	 
	public void draw(Graphics g) {
		
		if (displayImage != null) {
			
			/*
			 * Get the graphics and apply this objects transformations
			 * (rotation, etc.)
			 */
			Graphics2D g2d = (Graphics2D) g;
			applyTransformations(g2d);
			/* Actually draw the image, perform the pivot point translation here */
			g2d.drawImage(displayImage, 
					0, 
					0,
					(int) (getUnscaledWidth()),
					(int) (getUnscaledHeight()), null);
			
			/*
			 * undo the transformations so this doesn't affect other display
			 * objects
			 */
			reverseTransformations(g2d);
		}
	}

	/**
	 * Applies transformations for this display object to the given graphics
	 * object
	 * */
	protected void applyTransformations(Graphics2D g2d) {
		g2d.translate(this.position.getX() + this.pivotPoint.getX(), this.position.getY() + this.pivotPoint.getY());
		g2d.rotate(this.rotation*(3.1416)/180, this.pivotPoint.getX(), this.pivotPoint.getY());
		g2d.scale(this.scaleX, this.scaleY);
		g2d.setComposite(makeComposite(this.getAlpha()));
		if (this.getVisible()==false){
			g2d.setComposite(makeComposite(0));
		}
	}

	/**
	 * Reverses transformations for this display object to the given graphics
	 * object
	 * */
	protected void reverseTransformations(Graphics2D g2d) {
		g2d.setComposite(makeComposite(1));
		g2d.scale(1/this.scaleX, 1/this.scaleY);
		g2d.rotate(-(this.rotation*(3.1416)/180), this.pivotPoint.getX(), this.pivotPoint.getY());
		g2d.translate(-this.position.getX() - this.pivotPoint.getX(), -this.position.getY() - this.pivotPoint.getY());
	}

	public boolean getInPlay() {
		return inPlay;
	}

	public void setInPlay(boolean inPlay) {
		this.inPlay = inPlay;
	}

}